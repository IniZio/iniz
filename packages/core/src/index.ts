export { atom, isAtom } from "./atom";
export { effect, Effect } from "./effect";
export { batch } from "./batch";
