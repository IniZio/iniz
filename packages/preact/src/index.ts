export * from "@iniz/core";
export { useAtom } from "./useAtom";
export { useComputed } from "./useComputed";
export { useSideEffect } from "./useSideEffect";
export { useStore } from "./useStore";
import "./createElement";
