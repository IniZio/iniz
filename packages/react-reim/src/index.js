import React from 'react'

import createConsumer from './factories/consumer'
import createProvider from './factories/provider'

export function createContext(store) {
  const Context = React.createContext()

  const res = {
    get __isContext() {
      return true
    },
    Consumer: createConsumer(Context.Consumer),
    Provider: createProvider(Context.Provider, store)
  }

  Object.assign(store, res)
  return store
}

export function connect(store, getter = s => s, setter = () => ({})) {
  const Context = store.__isContext ? store : createContext(store)
  return Wrapped => p => (
    <Context.Consumer store={store} getter={getter} setter={setter}>
      {
        cache => <Wrapped {...cache} {...p}/>
      }
    </Context.Consumer>
  )
}
