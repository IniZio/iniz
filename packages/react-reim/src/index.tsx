import React, {PureComponent} from 'react'
import reim, {Store, Mutation, Getter} from 'reim'

import State, {StateProps} from './components/state'

export {default as State} from './components/state'

class ReactStore extends Store {
  get __isReactReim() {return true}
  Consumer: (props: StateProps) => React.ReactElement<StateProps>;
  get: (fn: Getter) => React.ReactElement<StateProps>;
}

// Syncs props to store
export function pipeTo(store: Store, mutation: Mutation = (_: any, p: any) => p): React.ComponentClass {
  class Piper extends PureComponent {
    componentDidUpdate() {
      store.set(mutation, this.props)
    }

    render = () => null
  }
  return Piper
}

export const react = () =>
  ({
    name: 'react',
    call(store: Store) {
      Object.defineProperties(store, {
        __isReactReim: {
          value: true
        },
        Consumer: {
          value: (props: any) => <State store={store} {...props}/>
        },
        get: {
          value: (fn: Getter) => <State store={store} getter={fn}/>
        }
      })
    }
  })

export function connect(store: ReactStore, getter: Getter = s => s, setter: Mutation = () => ({})) {
  const Context = store.__isReactReim ? store : store.plugin(react())
  return (Wrapped: React.ComponentClass) => (p: any) => (
    <Context.Consumer getter={getter} setter={setter}>
      {
        selected => <Wrapped {...selected} {...p}/>
      }
    </Context.Consumer>
  )
}

export function useReim(initial: any, getter: Getter = s => s, dependencies: string[] = []) {
  // @ts-ignore
  if (!React.useState) {
    throw new Error('React@16.7-alpha.2 is required to use Hooks')
  }

  const store = initial.__isReim ? initial : reim(initial)

  // @ts-ignore
  const {useState, useEffect, useRef} = React

  const mountRef = useRef()
  const [state, set] = useState(store.snapshot(getter))

  useEffect(() => {
    if (mountRef.current) {
      set(() => store.snapshot(getter))
    } else {
      mountRef.current = true
    }
  }, dependencies)

  useEffect(() => () => store.unsubscribe(set))

  store.subscribe(set, {getter})

  return [state, store.set]
}

export default react
