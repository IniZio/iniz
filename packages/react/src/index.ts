export * from "@reim/core";

export { useAtom } from "./useAtom";
export { useSideEffect } from "./useSideEffect";
export { useCompute } from "./useCompute";