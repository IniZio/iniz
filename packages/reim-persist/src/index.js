import merge from 'lodash/merge'

function canWriteStorage(storage) {
  try {
    storage.setItem('@@', 1);
    storage.removeItem('@@');
    return true;
  } catch (e) {}

  return false;
}

function getState(key, storage, value) {
  try {
    return (value = storage.getItem(key)) && typeof value !== 'undefined'
      ? JSON.parse(value)
      : undefined;
  } catch (err) {}

  return undefined;
}

function setState(key, state, storage) {
  return storage.setItem(key, JSON.stringify(state));
}

export default function persist(options = {}) {
  const storage = options.storage || (window && window.localStorage)
  const key = `reim/${options.key}`

  if (!canWriteStorage(storage)) {
    throw new Error('Invalid storage given')
  }

  return {
    name: 'persist',
    apply(store) {
      const saved = getState(key, storage)
      if (typeof saved === 'object' && saved !== null) {
        store.setState(merge({}, store.state, saved))
      }

      store.subscribe(state => {
        setState(key, state, storage)
      }, {...options.subscriber, immediate: true})

      return store
    }
  }
}
