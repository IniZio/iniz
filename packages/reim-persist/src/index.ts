import {merge} from 'lodash'

function canWriteStorage(storage) {
  try {
    storage.setItem('@@', 1)
    storage.removeItem('@@')
    return true
  } catch (error) {}

  return false
}

function snapshot(key, storage, value?: any) {
  try {
    value = storage.getItem(key)
    return typeof value === 'undefined' ? undefined : JSON.parse(value)
  } catch (error) {}

  return undefined
}

function set(key, state, storage) {
  return storage.setItem(key, JSON.stringify(state))
}

export default function persist(options: any = {}) {
  const storage =
    options.storage ||
    (typeof window === 'undefined' ? null : window && window.localStorage)

  if (!canWriteStorage(storage)) {
    throw new Error('Invalid storage given')
  }

  return (store): void => {
    // console.log('persist store? ', store)

    if (!store.name || store.name.length <= 0) {
      throw new Error(
        'You cannot persist an anonymous store, use `name` option in store'
      )
    }

    const key = `reim/${store.name}`
    const saved = snapshot(key, storage)
    if (typeof saved === 'object' && saved !== null) {
      store.reset(merge({}, store._state, saved))
    }

    store.subscribe(
      state => {
        set(key, state, storage)
      },
      {...options.subscriber, immediate: true}
    )
  }
}
