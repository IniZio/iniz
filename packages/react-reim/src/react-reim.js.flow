// @flow
import {Component} from 'react'

import {Mutation, Getter, Plugin, Store, State} from 'reim'

export type Setter = State | {(state: State):  void | State}

export function pipeTo(store: Store, mutation: Mutation): Component

export function context(): Plugin
export const createContext: Plugin

export const State: Component

export function connect(store: Store, getter: Getter): {(Wrapped: Component): Component}
