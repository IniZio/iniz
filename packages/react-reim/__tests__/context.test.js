/* eslint react/prop-types: 0 */
/* eslint-disable-next-line import/no-extraneous-dependencies  */
import renderer from 'react-test-renderer'
import React, {Component} from 'react'
import {register} from '../../reim/src'
import {createContext, connect, context} from '../src'

test('context returns Consumer and Provider', () => {
  const store = context(register)({yer: 43})

  expect(store.Consumer).toBeDefined()
  expect(store.Provider).toBeDefined()
})

test('Consumer should have change in store state reflected', () => {
  const store = createContext(register({yer: 43}))

  const component = renderer.create(
    <store.Provider>
      <store.Consumer>
        {
          state => (
            <div>
              <div id="value">{state.yer}</div>
            </div>
          )
        }
      </store.Consumer>
    </store.Provider>
  )
  store.setState(state => {
    state.yer += 88
  })
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Unmount Cunsumer should unsubscribe', () => {
  const store = createContext(register({yer: 43}))

  const getter = jest.fn()

  const component = renderer.create(
    <store.Provider>
      <store.Consumer getter={getter}>
        {
          state => (
            <div>
              <div id="value">{state.yer}</div>
            </div>
          )
        }
      </store.Consumer>
    </store.Provider>
  )
  store.setState(state => {
    state.yer += 88
  })
  expect(getter).toBeCalledTimes(2)
  component.unmount()
  store.setState(state => {
    state.yer *= 88
  })
  expect(getter).toBeCalledTimes(2)
})

test('Properties not included in getter should not trigger update', () => {
  const store = createContext(register({hel: 43, gee: 10}))

  const updated = jest.fn()

  class Listen extends Component {
    componentDidUpdate = updated

    render() {
      return <div>{this.props.hel}</div>
    }
  }

  renderer.create(
    <store.Provider>
      <store.Consumer getter={state => ({hel: state.hel})}>
        {
          state => (
            <Listen {...state}/>
          )
        }
      </store.Consumer>
    </store.Provider>
  )
  store.setState(state => {
    state.hel -= 22
  })
  expect(updated).toBeCalledTimes(1)
  store.setState(state => {
    state.gee += 88
  })
  expect(updated).toBeCalledTimes(1)
})

test('use convenience method connect', () => {
  const store = createContext(register({bom: 19}))

  const Container = connect(store, state => ({bom: state.bom}))(
    state => <div>{JSON.stringify(state)}</div>
  )

  const component = renderer.create(
    <store.Provider>
      <Container/>
    </store.Provider>
  )
  store.setState(state => {
    state.bom += 490
  })
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('change getter', () => {
  const store = createContext(register({dui: 12, geo: 'tie'}))

  class Getter extends Component {
    state = {
      getter: state => ({dui: state.dui})
    }

    render() {
      return (
        <store.Provider>
          <store.Consumer getter={this.state.getter}>
            {
              state => (
                <div>
                  <div id="selected">{JSON.stringify(state)}</div>
                </div>
              )
            }
          </store.Consumer>
        </store.Provider>
      )
    }
  }

  const component = renderer.create(<Getter/>)
  expect(component.toJSON()).toMatchSnapshot()

  // Change in getter should change passed in state
  const {instance} = component.root
  instance.setState({getter: s => s})
  expect(component.toJSON()).toMatchSnapshot()
})
