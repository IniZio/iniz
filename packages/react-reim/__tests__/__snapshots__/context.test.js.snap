// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Consumer should have change in store state reflected 1`] = `
<div>
  <div
    id="value"
  >
    131
  </div>
</div>
`;

exports[`change getter 1`] = `
<div>
  <div
    id="selected"
  >
    {"dui":12}
  </div>
</div>
`;

exports[`change getter 2`] = `
<div>
  <div
    id="selected"
  >
    {"dui":12,"geo":"tie"}
  </div>
</div>
`;

exports[`use convenience method connect 1`] = `
<div>
  {"bom":509}
</div>
`;
